File: goblin-dao-mongo/src/main/java/org/goblinframework/dao/mongo/bson/serializer/BsonObjectIdSerializer.java
package org.goblinframework.dao.mongo.bson.serializer;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import org.bson.types.ObjectId;
import org.goblinframework.dao.mongo.bson.GoblinBsonGenerator;

import java.io.IOException;

final public class BsonObjectIdSerializer extends JsonSerializer<ObjectId> {

  @Override
  public void serialize(ObjectId value, JsonGenerator gen, SerializerProvider serializers) throws IOException {
    if (value == null) {
      serializers.defaultSerializeNull(gen);
    } else {
      GoblinBsonGenerator.cast(gen).writeObjectId(value);
    }
  }
}

File: goblinframework-database-mongo/src/main/java/org/goblinframework/database/mongo/bson/serializer/BsonObjectIdSerializer.kt
package org.goblinframework.database.mongo.bson.serializer

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider
import org.bson.types.ObjectId
import org.goblinframework.database.mongo.bson.GoblinBsonGenerator

class BsonObjectIdSerializer : JsonSerializer<ObjectId>() {

  override fun serialize(value: ObjectId?, gen: JsonGenerator, serializers: SerializerProvider) {
    value?.run {
      trySerialize(this, gen)
    } ?: kotlin.run {
      serializers.defaultSerializeNull(gen)
    }
  }

  private fun trySerialize(value: ObjectId, gen: JsonGenerator) {
    (gen as? GoblinBsonGenerator)?.run {
      this.writeObjectId(value)
    } ?: throw UnsupportedOperationException()
  }
}
